### json 불러와서 저장하기

In [1]:
import numpy as np
import pandas as pd
import json
import jpype
from bs4 import BeautifulSoup as bs

data_path = "data"

json_train=[]
for i in range(0,18):
    json_file = json.loads(open("%s/%d.json" % (data_path, i)).read())
    json_train += json_file["articles"]

### json에서 title 형태소 추출

In [2]:
from konlpy.tag import Mecab

mecab = Mecab()

labeled_train = []

for cnt, article in enumerate(json_train):
    
    if cnt % 10000 == 0:
        print cnt

    labeled_train.append({
            "istroll": article["is_troll"],
            "title_pos": ["%s_%s" % (first, second) for first, second in mecab.pos(article["title"])]
        })

labeled_train = pd.DataFrame.from_dict(labeled_train)
labeled_train

0
10000
20000
30000
40000
50000
60000
70000
80000
90000
100000
110000
120000
130000
140000
150000
160000
170000


istroll                                       title_morphs
0        False      [일간_NNG, 워스트_NNP, 마을_NNG, 회관_NNG, 입니다_VCP+EC]
1        False        [일간_NNG, 워스트_NNP, 일_NNG, 워_NNG, 입니다_VCP+EC]
2        False                                       [오_IC, !_SF]
3        False                                      [1_SN, 등_NNB]
4        False            [여기_NP, 가_JKS, 수용소_NNG, 냐_VCP+EF, ?_SF]
5        False                   [여기_NP, 가_JKS, 일_NR, 워_IC, !_SF]
6        False                                      [지_MM, 젼_NNG]
7        False  [함수_NNG, 니_VCP+EC, +_SY, 빛돌_NNP, 이_JKS, 합동_NNG...
8        False  [캬_IC, ._SF, .._SY, 미루_VV, 고_EC, 미루_VV, 다_EC, ...
9        False            [일간_NNG, ㅋㅋㅋㅋㅋㅋㅋ워슼ㅋㅋㅋㅋㅋ틐ㅋㅋㅋㅋㅋㅋ_UNKNOWN]
10       False                [순위_NNG, 놀이_NNG, ㅁ_NNG, ㅇㅎ_UNKNOWN]
11       False               [민영화_NNG, 말_VV, 고_EC, 사영_NNG, 화_XSN]
12       False  [어떻게_MAG, 해야_VV+EC, 가장_MAG, 뛰어난_VA+ETM, 첫_MM, ...
13       False                      [경축_NNG, 세_MM, 가지_NNBC, !_SF]
14       False       [덕_NNG, 후판_NNG, 만들_VV, 어_EC, 주_VX, 세요_EP+EF]
15       False  [민영화_NNG, 누르_VV, 면_EC, 비_XPN, 추천_NNG, 했_XSV+EP...
16       False                                          [엉엉엉_MAG]
17       False          [기억_NNG, 이_JKS, 나_VV, 지_EC, 않_VX, 습니다_EF]
18       False                          [안녕_NNG, 하_XSV, 세요_EP+EF]
19       False               [아이디_NNG, 는_JX, 언제_MAG, 쓰_VV, 나요_EC]
20       False  [ㅇㅂ는_UNKNOWN, 애국_NNG, 보수_NNG, 면_VCP+EC, 여긴_NP+...
21       False           [어휴_IC, 워_IC, 충이_NNG, 들_XSN, ㅉㅉ_UNKNOWN]
22       False  [일_NR, 베_NNG, 편의점_NNG, 담배_NNG, 호빵_NNP, 사건_NNG,...
23       False  [오픈_NNG, 기념_NNG, 으로_JKB, 일_NNG, 워_NNG, 로고_NNG,...
24       False                                   [아야나미_NNP, !_SF]
25       False  [철도_NNG, 민영화_NNG, 한방_NNG, 에_JKB, 정리_NNG, 해준다_X...
26       False          [박정근_NNP, 만세_NNG, 해_VV+EC, 봐_VX+EF, ._SF]
27       False  [터키_NNP, 노조_NNG, 프랑스_NNP, 노조_NNG, 는_JX, 그럼_MAJ...
28       False  [모바일_NNP, 버전_NNG, 두_JX, 얼릉_MAG, 오픈_NNG, 하_VV, ...
29       False                                         [유루유리_NNP]
...        ...                                                ...
172014   False   [휴_IC, ._SF, .._SY, 우울_NNG, 하_XSV, 더라_EC, 농_NNG]
172015    True  [탈모_NNG, 에_JKB, 대해서_VV+EC, 질문_NNG, 입니다_VCP+EF,...
172016   False               [뜻밖_NNG, 의_JKG, 불꽃_NNG, 싸_VA, 다구_EC]
172017   False  [아_IC, 이마스_NNP, 보_VV, 는데_EC, 미키_NNG, 가_JKS, 잘_...
172018   False  [지금_MAG, 모바일_NNP, 사진_NNG, 지원_NNG, 안_MAG, 되_VV,...
172019   False  [어_IC, ?_SF, 저_MM, 트롤_NNG, 밭_NNG, 인가요_VCP+EF, ...
172020   False                                   [벌레_NNG, 출현_NNG]
172021   False      [저_IC, 왜_MAG, 트롤_NNG, 인가요_VCP+EF, ._SF, ._SF]
172022   False  [이명박_NNP, 의_JKG, 독특_NNG, 한_XSA+ETM, 영어_NNG, 인사...
172023   False  [농부_NNG, 들_XSN, 한테_JKB, 궁금_XR, 한_XSA+ETM, 게_NN...
172024   False  [본_MM, 농_NNG, 은_JX, 될_VV+ETM, 수_NNB, 있_VV, 음_E...
172025   False      [트롤_NNG, 밭_NNG, ._SF, ._SY, 무서웠_VA+EP, 어요_EF]
172026   False          [PS_SL, 4_SN, 비둘기_NNG, 연애_NNG, 시뮬레이션_NNG]
172027   False  [주연_NNG, 테크_NNP, 서_JKB, 나온_VV+ETM, 제이탭_NNP, s_...
172028   False                    [최근_NNG, 얼굴_NNG, 책_NNG, 에서_JKB]
172029   False                       [2014_SN, 년_NNB, ._SF, ._SF]
172030   False              [그러_VV, 고_EC, 보_VX, 니_EF, ._SF, ._SF]
172031   False  [농_NNG, 농_NNG, 들_XSN, 마우스_NNG, 추천_NNG, 해_XSV+E...
172032   False  [정규직_NNG, 과_JC, 비정규직_NNG, 이_JKS, 함께_MAG, 한_XSA...
172033   False      [어디_NP, 서_JKB, 본_VV+ETM, 듯_NNB, 한데_NNG, 2_SN]
172034   False  [(_SSO, 속보_NNG, )_SSC, 땅콩_NNG, 부사장_NNG, 님_XSN,...
172035   False               [농약_NNG, 충전_NNG, 용_XSN, ._SF, .._SY]
172036   False  [umc_SL, ,_SC, 이용_NNG, (_SSO, 뻘_NNG, 글_NNG, )_...
172037   False  [안녕_NNG, 하_XSV, 세요_EP+EF, ~_SY, 일_NNG, 워_NNG, ...
172038   False  [토토_NNP, 가_JKS, 열풍_NNG, ,_SC, 90_SN, 년_NNB, 대_...
172039   False  [몸_NNG, 이_JKS, 약해진_VA+EC+VX+ETM, 건지_NNB+VCP+EF...
172040   False  [살_VV, 면서_EC, 좋_VA, 은_ETM, 집_NNG, 주인_NNG, 은_JX...
172041   False  [C_SL, 9_SN, 외국_NNG, 섭_NNG, 하_VV, 시_EP, 는_ETM,...
172042   False  [끄_VV, 아_EC, 연말_NNG, 이_VCP, 라니_

###bag of words

In [3]:
from sklearn.feature_extraction.text import CountVectorizer

max_features = 1000

vectorizer = CountVectorizer(analyzer = "word", tokenizer = None, preprocessor = None,
                            stop_words = None, max_features=max_features)

train_data_features = vectorizer.fit_transform((labeled_train["title_pos"].apply(lambda x:" ".join(x)))).toarray()

col = ["bow_title_%s" % data for data in vectorizer.get_feature_names()]
labeled_train = pd.concat([labeled_train, pd.DataFrame(train_data_features, columns = col, index=labeled_train.index)],axis=1)

labeled_train

istroll                                       title_morphs  \
0        False      [일간_NNG, 워스트_NNP, 마을_NNG, 회관_NNG, 입니다_VCP+EC]   
1        False        [일간_NNG, 워스트_NNP, 일_NNG, 워_NNG, 입니다_VCP+EC]   
2        False                                       [오_IC, !_SF]   
3        False                                      [1_SN, 등_NNB]   
4        False            [여기_NP, 가_JKS, 수용소_NNG, 냐_VCP+EF, ?_SF]   
5        False                   [여기_NP, 가_JKS, 일_NR, 워_IC, !_SF]   
6        False                                      [지_MM, 젼_NNG]   
7        False  [함수_NNG, 니_VCP+EC, +_SY, 빛돌_NNP, 이_JKS, 합동_NNG...   
8        False  [캬_IC, ._SF, .._SY, 미루_VV, 고_EC, 미루_VV, 다_EC, ...   
9        False            [일간_NNG, ㅋㅋㅋㅋㅋㅋㅋ워슼ㅋㅋㅋㅋㅋ틐ㅋㅋㅋㅋㅋㅋ_UNKNOWN]   
10       False                [순위_NNG, 놀이_NNG, ㅁ_NNG, ㅇㅎ_UNKNOWN]   
11       False               [민영화_NNG, 말_VV, 고_EC, 사영_NNG, 화_XSN]   
12       False  [어떻게_MAG, 해야_VV+EC, 가장_MAG, 뛰어난_VA+ETM, 첫_MM, ...   
13       False                      [경축_NNG, 세_MM, 가지_NNBC, !_SF]   
14       False       [덕_NNG, 후판_NNG, 만들_VV, 어_EC, 주_VX, 세요_EP+EF]   
15       False  [민영화_NNG, 누르_VV, 면_EC, 비_XPN, 추천_NNG, 했_XSV+EP...   
16       False                                          [엉엉엉_MAG]   
17       False          [기억_NNG, 이_JKS, 나_VV, 지_EC, 않_VX, 습니다_EF]   
18       False                          [안녕_NNG, 하_XSV, 세요_EP+EF]   
19       False               [아이디_NNG, 는_JX, 언제_MAG, 쓰_VV, 나요_EC]   
20       False  [ㅇㅂ는_UNKNOWN, 애국_NNG, 보수_NNG, 면_VCP+EC, 여긴_NP+...   
21       False           [어휴_IC, 워_IC, 충이_NNG, 들_XSN, ㅉㅉ_UNKNOWN]   
22       False  [일_NR, 베_NNG, 편의점_NNG, 담배_NNG, 호빵_NNP, 사건_NNG,...   
23       False  [오픈_NNG, 기념_NNG, 으로_JKB, 일_NNG, 워_NNG, 로고_NNG,...   
24       False                                   [아야나미_NNP, !_SF]   
25       False  [철도_NNG, 민영화_NNG, 한방_NNG, 에_JKB, 정리_NNG, 해준다_X...   
26       False          [박정근_NNP, 만세_NNG, 해_VV+EC, 봐_VX+EF, ._SF]   
27       False  [터키_NNP, 노조_NNG, 프랑스_NNP, 노조_NNG, 는_JX, 그럼_MAJ...   
28       False  [모바일_NNP, 버전_NNG, 두_JX, 얼릉_MAG, 오픈_NNG, 하_VV, ...   
29       False                                         [유루유리_NNP]   
...        ...                                                ...   
172014   False   [휴_IC, ._SF, .._SY, 우울_NNG, 하_XSV, 더라_EC, 농_NNG]   
172015    True  [탈모_NNG, 에_JKB, 대해서_VV+EC, 질문_NNG, 입니다_VCP+EF,...   
172016   False               [뜻밖_NNG, 의_JKG, 불꽃_NNG, 싸_VA, 다구_EC]   
172017   False  [아_IC, 이마스_NNP, 보_VV, 는데_EC, 미키_NNG, 가_JKS, 잘_...   
172018   False  [지금_MAG, 모바일_NNP, 사진_NNG, 지원_NNG, 안_MAG, 되_VV,...   
172019   False  [어_IC, ?_SF, 저_MM, 트롤_NNG, 밭_NNG, 인가요_VCP+EF, ...   
172020   False                                   [벌레_NNG, 출현_NNG]   
172021   False      [저_IC, 왜_MAG, 트롤_NNG, 인가요_VCP+EF, ._SF, ._SF]   
172022   False  [이명박_NNP, 의_JKG, 독특_NNG, 한_XSA+ETM, 영어_NNG, 인사...   
172023   False  [농부_NNG, 들_XSN, 한테_JKB, 궁금_XR, 한_XSA+ETM, 게_NN...   
172024   False  [본_MM, 농_NNG, 은_JX, 될_VV+ETM, 수_NNB, 있_VV, 음_E...   
172025   False      [트롤_NNG, 밭_NNG, ._SF, ._SY, 무서웠_VA+EP, 어요_EF]   
172026   False          [PS_SL, 4_SN, 비둘기_NNG, 연애_NNG, 시뮬레이션_NNG]   
172027   False  [주연_NNG, 테크_NNP, 서_JKB, 나온_VV+ETM, 제이탭_NNP, s_...   
172028   False                    [최근_NNG, 얼굴_NNG, 책_NNG, 에서_JKB]   
172029   False                       [2014_SN, 년_NNB, ._SF, ._SF]   
172030   False              [그러_VV, 고_EC, 보_VX, 니_EF, ._SF, ._SF]   
172031   False  [농_NNG, 농_NNG, 들_XSN, 마우스_NNG, 추천_NNG, 해_XSV+E...   
172032   False  [정규직_NNG, 과_JC, 비정규직_NNG, 이_JKS, 함께_MAG, 한_XSA...   
172033   False      [어디_NP, 서_JKB, 본_VV+ETM, 듯_NNB, 한데_NNG, 2_SN]   
172034   False  [(_SSO, 속보_NNG, )_SSC, 땅콩_NNG, 부사장_NNG, 님_XSN,...   
172035   False               [농약_NNG, 충전_NNG, 용_XSN, ._SF, .._SY]   
172036   False  [umc_SL, ,_SC, 이용_NNG, (_SSO, 뻘_NNG, 글_NNG, )_...   
172037   False  [안녕_NNG, 하_XSV, 세요_EP+EF, ~_SY, 일_NNG, 워_NNG, ...   
172038   False  [토토_NNP, 가_JKS, 열풍_NNG, ,_SC, 90_SN, 년_NNB, 대_...   
172039   False  [몸_NNG, 이_JKS, 약해진_VA+EC+VX+ETM, 건지_NNB+VCP+EF...   
17204

### predictor, model 생성

In [4]:
from sklearn.ensemble import RandomForestClassifier

label = 'istroll'
pre = labeled_train.columns.drop(['title_pos', label])

model = RandomForestClassifier(n_estimators=10, n_jobs=6)

###cross validation

In [5]:
from sklearn import cross_validation

cv_value = 3

scores = cross_validation.cross_val_score(model, labeled_train[pre], labeled_train[label], cv=cv_value, scoring="roc_auc")
cv_result = scores.mean()

print(cv_result)

0.630053006037


##Result
  * title의 morphs를 bag of words로 변환 (feature : 1,000개) - 0.612752193303
  * title의 morphs를 bag of words로 변환 (feature : 1,000개) - 0.630053006037

In [6]:
import pickle

pickle.dump(labeled_train[pre], open("title_%d.p" % max_features, "wb"), protocol = pickle.HIGHEST_PROTOCOL)

bow_title_017  bow_title_10  bow_title_100  bow_title_11  \
0                   0             0              0             0   
1                   0             0              0             0   
2                   0             0              0             0   
3                   0             0              0             0   
4                   0             0              0             0   
5                   0             0              0             0   
6                   0             0              0             0   
7                   0             0              0             0   
8                   0             0              0             0   
9                   0             0              0             0   
10                  0             0              0             0   
11                  0             0              0             0   
12                  0             0              0             0   
13                  0             0              0             0   
14                  0             0              0             0   
15                  0             0              0             0   
16                  0             0              0             0   
17                  0             0              0             0   
18                  0             0              0             0   
19                  0             0              0             0   
20                  0             0              0             0   
21                  0             0              0             0   
22                  0             0              0             0   
23                  0             0              0             0   
24                  0             0              0             0   
25                  0             0              0             0   
26                  0             0              0             0   
27                  0             0              0             0   
28                  0             0              0             0   
29                  0             0              0             0   
...               ...           ...            ...           ...   
172014              0             0              0             0   
172015              0             0              0             0   
172016              0             0              0             0   
172017              0             0              0             0   
172018              0             0              0             0   
172019              0             0              0             0   
172020              0             0              0             0   
172021              0             0              0             0   
172022              0             0              0             0   
172023              0             0              0             0   
172024              0             0              0             0   
172025              0             0              0             0   
172026              0             0              0             0   
172027              0             0              0             0   
172028              0             0              0             0   
172029              0             0              0             0   
172030              0             0              0             0   
172031              0             0              0             0   
172032              0             0              0             0   
172033              0             0              0             0   
172034              0             0              0             0   
172035              0             0              0             0   
172036              0             0              0             0   
172037              0             0              0             0   
172038              0             0              0             0   
172039              0             0              0             0   
172040              0             0              0             0